Add sorting to lists

6a68 opened the pull request:
Cleaned up and ready to land.

Fixes #35, #93.

## Testing and Review Notes

Adds management page UI to sort the list by name, last changed, or last used. Sorting should work while filtering. Sort choice should persist across manage page reloads.

Popup entries should always be displayed in last used order.

## Screenshots or Videos

_(Optional: to clearly demonstrate the feature or fix to help with testing and reviews)_


## To Do

- add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
  - optional: consider adding integration tests
- [x] request the "UX" team perform a design review (if/when applicable)
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- [x] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-addon/developer/test-plan-accessibility/) of any added UI

linuxwolf commented on src/list/manage/components/item-list-panel.js:
```diff
@@ -38,6 +39,7 @@ export default function ItemListPanel({className, inputRef, totalItemCount,
         <ItemFilter className={styles.itemFilter}
                     inputRef={inputRef}/>
         <AddItem/>
+        <ListSort sort={sort} className={styles.listSort} {...props} />
```
This layout puts everything into a single row.  Maybe we should consider some options:
* moving `<ListSort/>` (or a wrapper of it?) up a level to be a sibling of `<PanelHeader/>` and `<PanelBody/>`; or
* making `<PanelHeader/>`'s orientation configurable (flex column vs flex row)

I fear trying to make `<PanelHeader/>` a grid is more trouble than it's worth ...

linuxwolf commented on src/widgets/select.js:
```diff
@@ -0,0 +1,69 @@
+  render() {
+    const { className, options, disabled, ...props } = this.props;
+    return (
+      <select value={this.state.value}
```
From past experience, trying to style `<select/>` to the degree our designs call for has been very problematic (maybe impossible).  I think this `<Select/>` component is a good abstraction over what we might ultimately do, though it's worth taking a bit to see if such a component exists out in the larger ecosystem.

lmorchard commented on src/list/manage/components/item-list-panel.js:
```diff
@@ -35,9 +37,16 @@ export default function ItemListPanel({className, inputRef, totalItemCount,
+          <AddItem/>
+        </div>
+        <div className={styles.secondRow}>
+          <ListSort sort={sort} className={styles.listSort} {...props} />
```
Are ListSort or ListCounter used anywhere else separately? Might reduce complexity a little to combine this whole row into a single component? (Well, you know, other than the busy work to combine them, changes tests, etc 😅 )

6a68 replied:
Good questions! I think the ListCounter could abstract away the "foo entries found" code that's [currently inlined](https://github.com/mozilla-lockbox/lockbox-addon/blob/master/src/list/popup/components/item-list-panel.js#L72-L92) in the popup's ItemListPanel, though it'd be a bit of work to add a non-numeric default state to show the "recently used entries. click to search" text when there's no filter.

lmorchard commented on src/list/manage/components/list-sort.js:
```diff
@@ -0,0 +1,71 @@
+import styles from "./list-sort.css";
+
+export default class ListSort extends React.Component {
+  // TODO: do we need to do something with className?
```
Probably not, unless it's expected that someone using this will want to select a style variant. Could be YAGNI for now.

lmorchard commented on src/list/manage/components/list-sort.js:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  render() {
+    const { onChange, sort, className, disabled, ...props } = this.props;
```
Per earlier comment that onChange is confusing, I don't think you need onChange destructured here 😁 Same for className

lmorchard commented on src/list/manage/index.js:
```diff
@@ -15,22 +15,29 @@ import reducer from "./reducers";
-initializeMessagePorts(store);
-store.dispatch(getProfile());
+let store;
+(async () => {
```
Hmm, this *should* be near-immediate and not require any kind of visual loading indicator. But, wrapping the entire initialization in an async block waiting on the sort loading seems like it could be trouble.

Could you initialize the app, *then* start the async load and dispatch an action to update the state? (I guess that might cause a flash of content shifting around.) Also, is there a chance that an exception could get thrown here?

6a68 replied:
More good questions! Re-sorting the list is going to be a highly visible change, so it seemed better to force it to wait a beat to hit localstorage.

There is indeed exception handling to deal with, but that happens in sort-middleware.js, in which case the `loadSort` return value is falsy, and we'll wind up going with the default sort.

6a68 commented:
Cleaned up all the unit test bustage due to changing the item summary and adding the sort state to the redux cache, and added tests for the counter and the sort utils. Just need to get the ListSort tests filled in, take a second look at the async init issue Les pointed out, and make sure I've hit all the other points from past reviews. I'll pluck the 'WIP' off this when it's good for a final pass.

6a68 commented:
Whoops, not quite ready to land yet, looks like I've got a minor CSS bug:

<img width="352" alt="Screen Shot 2019-03-14 at 9 26 10 AM" src="https://user-images.githubusercontent.com/96396/54373893-45d34800-463b-11e9-8fb3-e76fc13110e2.png">

6a68 commented:
Hmm. @linuxwolf do you know if we have any mocks for a ghost town / no entries state for the doorhanger? Not quite sure what that _should_ look like

linuxwolf commented:
> Hmm. @linuxwolf do you know if we have any mocks for a ghost town / no entries state for the doorhanger? Not quite sure what that _should_ look like

The `test/unit/list/manage/mock-redux-state.js` exports an `initialState` with no entries.  Should pull that up a couple levels, eventually.

6a68 commented:
@linuxwolf heh, I meant UX mocks, not testing mocks ;-)

linuxwolf commented:
> @linuxwolf heh, I meant UX mocks, not testing mocks ;-)

🤦‍♂️ sorry!  The closest we have are "no matching entries"

* [invision](https://mozilla.invisionapp.com/share/ENOL84959AK#/screens/342865705)
* [Zeplin](https://zpl.io/agP3rj9)

6a68 commented:
Fixed the last little CSS nit, should be ready to land

linuxwolf commented on karma.conf.js:
```diff
@@ -30,6 +30,9 @@ module.exports = (config) => {
         { type: "html", subdir: "html" },
       ],
     },
+    mochaReporter: {
```
noice

linuxwolf commented:
In running this branch, I noticed things get out of shape if the viewport width is small enough:
<img width="254" alt="image" src="https://user-images.githubusercontent.com/757401/54392315-f279ed80-466c-11e9-8285-033bf71684e6.png">

Looking a little deeper, it appears the left-side list doeesn't have a minimum width.  Do we want to try and tackle that as part of this PR, or file a follow up?  I lean a little more toward a follow up, because I think other things might be affected (e.g., #99).

linuxwolf commented on src/list/manage/components/item-list-panel.css:
```diff
@@ -2,17 +2,51 @@
 }
+
+.list-sort {
+}
```
This is present because of the imported CSS is expecting it, right?  If so, we probably want a comment so future us doesn't mistakenly remove it.

6a68 replied:
I think we can just get rid of it

linuxwolf commented on src/icons/arrow-dropdown.svg:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" ?><svg height="48" viewBox="0 0 48 48" width="48" xmlns="http://www.w3.org/2000/svg"><path d="M14 20l10 10 10-10z"/><path d="M0 0h48v48h-48z" fill="none"/></svg>
```
The rest of our SVG icons are internally sized correctly, but it wasn't working out initially for this one.  I think we can change the `width` and `height` here, and maybe ditch setting the dimensions in CSS.

linuxwolf commented on test/unit/background/message-ports-test.js:
```diff
@@ -115,6 +127,15 @@ describe("background > message ports", () => {
+    expect(result.item.timePasswordChanged < Date.now());
+    expect(result.item.timePasswordChanged > originalTime);
+
+    // Now, delete timePasswordChanged everywhere, so the items will be
```
😞  but understandable

6a68 commented:
@linuxwolf Thanks for the final pass! :beers:

Looks like `list/manage/components/app.css` does have a min-width of 150px on the `aside` class. Cranking it up to 250px fixes the collapsing list at narrow sizes, though the rest of the page still needs some min-width help:

<img width="466" alt="Screen Shot 2019-03-14 at 4 00 28 PM" src="https://user-images.githubusercontent.com/96396/54397240-058fbc00-4673-11e9-8de8-84a5f904a947.png">

Fixed the other nits you pointed out. Merging.
